### DISCUSSION TITLE
Basket Transactions and Awards

### DISCUSSION THREAD

**Discussion by tuyilmaz** [YimMenu/HorseMenu#192] [Ideas]
I've decided to share what I've found about transactions, since I don't have enough time to work on it. 

`NETSHOPPING::_0xA3B8D31C13CB4239` is used for moonshine and trader production, buying/selling items from/to stores, claiming awards (including serverside ones), etc. For example while claiming a 50 RDO$ voucher this native is called like this:

First we need to start the transaction:
```
int TransactionID = -1;
NETSHOPPING::_CASHINVENTORY_TRANSACTION_START(&TransactionID, "basket"_J, "BUYAWARD"_J);
```
Here is the arguments for `_0xA3B8D31C13CB4239`:
1. TransactionID: The id we get after calling `_CASHINVENTORY_TRANSACTION_START`.
2. ActionHash: [Cashinventory transaction action hash](https://github.com/femga/rdr3_discoveries/blob/master/AI/EVENTS/cashinventory_transition_actions.lua) In this case 541670136 (BUYAWARD).
3. ItemInfo: Pointer to a struct containing information of the item that'll be used in transaction.
4. ItemInfoCount: Item count of ItemInfo.
5. ItemInfo2: Another struct pointer. It is the same with the 3rd arg in this case.
6. unk: 0 in most cases.

Here is the ItemInfo in our case:
```
uint64_t ItemInfo[18];
ItemInfo[8] = 1905403949; //Award Hash
ItemInfo[9] = 1; //how many times to give the award(uncertain)
ItemInfo[10] = 1; //unk
ItemInfo[11] = 2703302912; //constant
ItemInfo[12] = 1248274121; //constant
ItemInfo[13] = 205718222; //acquirecosts item key, which is the key of the item you must have to claim the award. voucher in this case
ItemInfo[17] = 1905403949; //Award Hash again
```

After creating this struct we call the native:
`NETSHOPPING::_0xA3B8D31C13CB4239(TransactionID, "BUYAWARD"_J, &ItemInfo, 18, &ItemInfo, 0);`

Lastly we need to checkout:
`NETSHOPPING::_CASHINVENTORY_TRANSACTION_CHECKOUT(TransactionID);`

If we did everything right and have the voucher we'll get 50 RDO$.

**Comment by Rxann** [comment]
Absolutely phenomenal work. Thanks for sharing. I will look at this more when I get home.

**Comment by tuyilmaz** [comment]
I've found something strange. It is possible to level up an expired season! We can complete season_006(idk it's name) by claiming 0x755AA556 award via GiveItemDatabaseAward script function. This is the only season that has a clientside award, but we still get its free tier awards after claiming this award, which means it can be possible to unlock previous season awards.
By the way catalog_awards_mp.ymt doesn't contain all the awards available for example regular xp, honor, dead eye, health, stamina awards.

Apart from these, @Rxann do we have the class for rage::atMap.

**Reply by Rxann** [reply]
> I've found something strange. It is possible to level up an expired season! We can complete season_006(idk it's name) by claiming 0x755AA556 award via GiveItemDatabaseAward script function. This is the only season that has a clientside award, but we still get its free tier awards after claiming this award, which means it can be possible to unlock previous season awards.
> 
> By the way catalog_awards_mp.ymt doesn't contain all the awards available for example regular xp, honor, dead eye, health, stamina awards.
> 
> 
> 
> Apart from these, @Rxann do we have the class for rage::atMap.

This is good news. Very cool discovery. Sadly, we do not have atMap as of right now.

**Reply by tuyilmaz** [reply]
Since seasons have the same system with role levels, I'm going to reset my char and continue my research. If I can't find a way all my limited time items will be gone. Wish me luck :)

**Reply by Rxann** [reply]
> Since seasons have the same system with role levels, I'm going to reset my char and continue my research. If I can't find a way all my limited time items will be gone. Wish me luck :)

Alright, thank you for doing this, and good luck.

**Reply by tuyilmaz** [reply]
@Rxann giving awards sends "GAMESERVER_TXN_AWARDS" metric. Adding it to bad_metrics would be great.

**Reply by Rxann** [reply]
> @Rxann giving awards sends "GAMESERVER_TXN_AWARDS" metric. Adding it to bad_metrics would be great.

Okay, I have this drafted to block it. Thanks for this though.

**Comment by tuyilmaz** [comment]
Here is a simplified list of all clientside awards: [link](https://github.com/tuyilmaz/RDR3-Research/blob/6b9e6f0a3f050ce7062172e7d0a6a36f337aff82/mp_awards_clientside.json)

**Reply by Rxann** [reply]
> Here is a simplified list of all clientside awards: [link](https://github.com/tuyilmaz/RDR3-Research/blob/6b9e6f0a3f050ce7062172e7d0a6a36f337aff82/mp_awards_clientside.json)

Alright thanks, let me check it out.

**Reply by Rxann** [reply]
@tuyilmaz 
This seems to be the key to adding gold right here: `0x012118D3`
correct me if im wrong

EDIT:
It looks like a lot of things award these gold bars actually, I might have been mistaken.

**Reply by Rxann** [reply]
This `0xB2220066` looks to have unlimited claims to 100 gold bars

**Reply by tuyilmaz** [reply]
> This `0xB2220066` looks to have unlimited claims to 100 gold bars

Actually as you know gold bars have nuggets, so the last two digits are nuggets. Real gold value is that amount divided by 100

**Reply by Rxann** [reply]
> > This `0xB2220066` looks to have unlimited claims to 100 gold bars
> 
> Actually as you know gold bars have nuggets, so the last two digits are nuggets. Real gold value is that amount divided by 100

Ah okay, I was afraid of this. Looks like I will need to find a bigger amount.

**Reply by tuyilmaz** [reply]
> > > This `0xB2220066` looks to have unlimited claims to 100 gold bars
> > 
> > Actually as you know gold bars have nuggets, so the last two digits are nuggets. Real gold value is that amount divided by 100
> 
> Ah okay, I was afraid of this. Looks like I will need to find a bigger amount.

100 is the biggest one clientside

**Reply by Rxann** [reply]
> 100 is the biggest one clientside

Ok and since we are limited to clientside, I guess this would have to do for the gold adder. Is it possible to add multiple quantities to you basket before you checkout?

EDIT:
I reviewed the original post that was posted, and it looks like it is possible.

**Reply by tuyilmaz** [reply]
I don't know if clientside awards can be claimed with basket transactions. They have a different method in game scripts. You can look at the GiveItemDatabaseAward script function we use, to understand it better.

**Reply by Rxann** [reply]
@tuyilmaz 
Is it possible to trigger a transaction without an acquirecosts item key? Also, do you have a list of those?

**Reply by tuyilmaz** [reply]
@Rxann 
Catalog awards can be found [here](https://raw.githubusercontent.com/tuyilmaz/RDR3-Research/main/catalog_awards_mp.xml). Currently we can only claim the ones where serveronly and debugonly are false. Serveronly awards use basket transactions. I have no idea if it is possible to claim debugonly ones.
For acquirecosts item keys, you can found them in [catalog_mp](https://raw.githubusercontent.com/tuyilmaz/RDR3-Research/main/catalog_mp.xml). Just search for the award hash and you'll see acquirecosts with item(s) in it. I've tried doing the transaction with acquirecosts item key = 0, but it didn't work.

**Reply by tuyilmaz** [reply]
@Rxann 
Do you know anything about how SECURE_LOAD and SECURE_STORE functions work?

**Reply by Rxann** [reply]
> @Rxann Do you know anything about how SECURE_LOAD and SECURE_STORE functions work?

No, I didn't even know these existed tbh. I don't see them in the [NativeDB](https://alloc8or.re/rdr3/nativedb/)

**Reply by tuyilmaz** [reply]
> No, I didn't even know these existed tbh. I don't see them in the [NativeDB](https://alloc8or.re/rdr3/nativedb/)

They're scriptvm operations, not native functions. If you have the decompiled scripts, search for SECURE_STORE and you'll see it.

**Reply by Rxann** [reply]
> They're scriptvm operations, not native functions. If you have the decompiled scripts, search for SECURE_STORE and you'll see it.

Ah okay, I see this now. Yeah I still have not a clue on the functionality of this. Could be something that is potentially very interesting.

**Reply by tuyilmaz** [reply]
@maybegreat48 
Since you updated the script decompiler to work with RDR 2. Do you know how these work?
